game.py

```diff
@@ -0,0 +1,302 @@
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = self._player_next
+        previous_game_state = deepcopy(self)
+        return GameState(board, player_next, previous_game_state)
+
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def make_move(self, move):
+        new_game_state = self._copy()
+        new_game_state._board[move[0]][move[1]] = new_game_state._player_next
+        new_game_state._player_next = - new_game_state._player_next
+        return new_game_state
+        
+    def unmake_move(self):
+        new_game_state = self._previous_game_state
+        return new_game_state
+        
+    def get_next_moves(self, moves_sort = False, game_end = False):
+        moves = []
+        any_moves = False
+        for i in range(len(self._board)):
+            if any_moves:
+                break
+            for j in range(len(self._board[i])):
+                if self._board[i][j] == None:
+        #            moves.append ((i, j))
+                    moves.insert(0, (i, j))
+                    if game_end:
+                        any_moves = True
+                        break
+                    
+        random.shuffle(moves)
+        
+        if moves_sort:
+            moves_and_scores = []
+            i = 0
+            for move in moves:
+                new_game_state = self.make_move(move)
+                moves_and_scores.append([move])
+                moves_and_scores[i].append(new_game_state.score())
+                i = i + 1
+            print ('Moves and scores before sort : ', moves_and_scores)
+            moves_and_scores.sort(key = lambda x: x[1])
+            print ('Moves and scores after sort : ', moves_and_scores)
+            moves = []
+            for move_and_score in moves_and_scores:
+                moves.append(move_and_score[0])
+                
+        # print('moves are ', moves)
+        return moves
+
+    
+
+
+    def game_end(self):
+        if self.get_next_moves(game_end = True) == []:
+            return True
+        else:
+            return False
+
+    def valid_move(self, move):
+        if move in self.get_next_moves():
+            return True
+        else:
+            return False
+   
+        
+    def score(self):
+        score = 0
+        for i in range(len(self._board)):
+            for j in range(len(self._board[i])):
+                if self._board[i][j] == self._player_next:
+                    score = score + (i + 1)*(j + 1)
+                else:
+                    if self._board[i][j] == - self._player_next:
+                        score = score - (i + 1)*(j + 1)
+
+        global score_count
+        score_count = score_count + 1
+
+        ## print('score = ', score)
+        return score
+
+def minimax(game_state, look_forward,  player_next_first, alpha, beta, sort_depth):

```

Comment:
minimax() works as intended, picking optimal moves for this game, but it's interesting to see it work like this on such a simple game. Right now the algorithm assigns no value to who is next to play, we just use the end-game-scoring mechanism secondarily as a stand-in heuristic for mid-game-evaluation.

So right now it never makes a wrong move even with 1-look-forward, because this is a game where greedy moves are always best, but it is often quite wrong about what it thinks the value of the game is.

Excited to see this develop!